apps/obsidian/src/components/SearchBar.tsx

```diff
@@ -0,0 +1,224 @@
+import { App, debounce } from "obsidian";
+import React, { useEffect, useState, useMemo, ReactNode } from "react";
+import { QueryEngine } from "~/services/QueryEngine";
+
+type SearchBarBaseProps<T> = {
+  app: App;
+  placeholder?: string;
+  onNodeSelect: (item: T | null) => void;
+  debounceMs?: number;
+  minQueryLength?: number;
+  getItemText: (item: T) => string;
+  getItemKey: (item: T) => string;
+  renderItemContent?: (item: T, isSelected?: boolean) => ReactNode;
+  itemStyle?: React.CSSProperties;
+  containerStyle?: React.CSSProperties;
+  inputStyle?: React.CSSProperties;
+  resultsContainerStyle?: React.CSSProperties;
+};
+
+type OptionsSearchBarProps<T> = SearchBarBaseProps<T> & {
+  options: T[];
+  searchFunction?: (query: string) => Promise<T[]>;
+};
+
+type FunctionSearchBarProps<T> = SearchBarBaseProps<T> & {
+  options?: never;
+  searchFunction: (query: string) => Promise<T[]>;
+};
+
+export type SearchBarProps<T> =
+  | OptionsSearchBarProps<T>
+  | FunctionSearchBarProps<T>;
+
+export function SearchBar<T extends {}>({
```

Comment:
Are there any Obsidian native classes we can take advantage of here instead of creating our own?  Maybe something like PopoverSuggest or AbstractInputSuggest or similar?

Reply:
yes, updated the SearchBar using AbastractInputSuggest